source/common/router/config_impl.cc

```diff
@@ -2147,13 +2146,42 @@ PerFilterConfigs::PerFilterConfigs(
     const Protobuf::Map<std::string, ProtobufWkt::Any>& typed_configs,
     const OptionalHttpFilters& optional_http_filters,
     Server::Configuration::ServerFactoryContext& factory_context,
-    ProtobufMessage::ValidationVisitor& validator) {
-  for (const auto& it : typed_configs) {
-    const auto& name = it.first;
-    auto object = createRouteSpecificFilterConfig(name, it.second, optional_http_filters,
-                                                  factory_context, validator);
-    if (object != nullptr) {
-      configs_[name] = std::move(object);
+    ProtobufMessage::ValidationVisitor& validator)
+    : ignore_optional_option_from_hcm_for_route_config_(Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.ignore_optional_option_from_hcm_for_route_config")) {
+
+  static const std::string filter_config_type =
+      envoy::config::route::v3::FilterConfig::default_instance().GetDescriptor()->full_name();
+
+  for (const auto& per_filter_config : typed_configs) {
+    const std::string& name = per_filter_config.first;
+    RouteSpecificFilterConfigConstSharedPtr config;
+
+    // Ignore the optional_http_filters if the ignore_optional_option_from_hcm_for_route_config_
+    // is set to true by the runtime feature.
+    bool is_optional_by_hcm = !ignore_optional_option_from_hcm_for_route_config_ &&
+                              (optional_http_filters.find(name) != optional_http_filters.end());
+
+    if (TypeUtil::typeUrlToDescriptorFullName(per_filter_config.second.type_url()) ==
+        filter_config_type) {
+      envoy::config::route::v3::FilterConfig filter_config;
+      Envoy::Config::Utility::translateOpaqueConfig(per_filter_config.second, validator,
+                                                    filter_config);
+
+      if (!filter_config.has_config()) {
+        continue;
+      }
+
+      config = createRouteSpecificFilterConfig(name, filter_config.config(),
```

Comment:
hmmm, is there some understand gaps? 

The `is_optional` flag in the FilterConfig of route is not used to override the `is_optional` flag in the HCM. They are two completely independent flags.

The one in the HCM will tell envoy a HTTP filter is optional when envoy loading the HCM. The one in the FilterConfig of route will tell envoy a HTTP filter ('s route configuration) is optional when loading the RouteConfiguration. 

I know in some way they are duplicate. But we have no better choice. We know nothing about the HCM's flags when we loading the RouteConfiguration, so additional/duplicated `is_optional` flag is necessary.

Reply:
I think my issue is that AIUI, the `FilterConfig.config` can be of a filter type that is completely different than the HCM filter type. I understand this should not be encouraged, but I think this is possible.
Then the question is what should happen if the HCM filter is valid, but the `FilterConfig.config` points to a config that is unknown and optional. I think Mark's [comment](https://github.com/envoyproxy/envoy/pull/27263#issuecomment-1555096822) mentions this case:
>  * If the client supports the RBAC filter but not the HTTPFault filter, then it will configure the RBAC filter in the LDS resource, but it will not use the HTTPFault entry in the RDS resource, so there will be no conflict when putting the two resources together.

It is a bit unclear whether the RBAC filter should be used on the data-path when the route is used or not, and I just think the `FilterConfig` comments could be improved in this case.

Reply:
Forget `is_optional` for a moment.  How would you want to handle this if the data plane supports both filters?  I think we have basically two choices:

1. Fail requests at L7.
2. Ignore the `typed_per_filter_config` entry that's for the wrong filter type and use only the top-level config.

If we think that the answer in this case is (2), then `is_optional` doesn't matter, because we'd have exactly the same behavior regardless of whether the filter in the `typed_per_filter_config` entry was supported.

If we think that the answer is (1), then that might be an argument in favor of doing the same thing in the `is_optional` case.  But we'd have to actually do some additional bookkeeping to make this work: when we validate `typed_per_filter_config`, if we see an unsupported filter that's marked as `is_optional`, instead of skipping it completely, we'd need to record that the entry as an invalid config, so that when we put the LDS and RDS resources together, the RBAC filter can tell that the override config is invalid and can fail requests.

Personally, I think that the right answer for the case where both filters are supported is actually (1), which is what gRPC does.  However, for the case with an unsupported filter marked as `is_optional`, I don't think that it's worth the additional bookkeeping to make sure we have the same behavior; I'm fine with allowing that case to just ignore the override config.

Reply:
> I think my issue is that AIUI, the FilterConfig.config can be of a filter type that is completely different than the HCM filter type. I understand this should not be encouraged, but I think this is possible.

Yes. And it almost alyways different. Like `envoy.extensions.filters.http.ext_authz.v3.ExtAuthz` and `envoy.extensions.filters.http.ext_authz.v3.ExtAuthzPerRoute`, `envoy.extensions.filters.http.lua.v3.Lua` and `envoy.extensions.filters.http.lua.v3.LuaPerRoute`.

> Then the question is what should happen if the HCM filter is valid, but the FilterConfig.config points to a config that is unknown and optional.

The route level per filter config will be nullptr, but the filter will be executed.

Reply:
See my comment above about `disabled`.  But I think that's mostly a separate issue; that problem already exists, since gRPC does have behavior (1) here.

For the current PR, I would argue that Envoy should also have behavior (1).  That way, when we eventually build a mechanism for clients to report the kind of config failures that are visible only when combining multiple resources, this will be considered an error and can be reported accordingly.

Reply:
Agree this should be a independent issue. I will mark this resolved here to make the context of this PR more clean.